build.go

```diff
@@ -68,6 +71,139 @@ func if_not_exists(cmd string, cb func()) {
+	r, err := zip.OpenReader(src)
+    if err != nil {
+        return err
+    }
+    defer func() {
+        if err := r.Close(); err != nil {
+            panic(err)
+        }
+    }()
+
+    os.MkdirAll(dest, 0755)
+
+    // Closure to address file descriptors issue with all the deferred .Close() methods
+    extractAndWriteFile := func(f *zip.File) error {
+       rc, err := f.Open()
+        if err != nil {
+            return err
+        }
+        defer func() {
+            if err := rc.Close(); err != nil {
+                panic(err)
+            }
+        }()
+
+        path := filepath.Join(dest, f.Name)
+
+        if f.FileInfo().IsDir() {
+            os.MkdirAll(path, f.Mode())
+        } else {
+            os.MkdirAll(filepath.Dir(path), f.Mode())
+            f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
+            if err != nil {
+                return err
+            }
+            defer func() {
+                if err := f.Close(); err != nil {
+                    panic(err)
+                }
+            }()
+
+            _, err = io.Copy(f, rc)
+            if err != nil {
+                return err
+            }
+        }
+        return nil
+    }
+
+    for _, f := range r.File {
+			if strings.Contains((*f).Name, fileToExtract) {
+        err := extractAndWriteFile(f)
+        if err != nil {
+					return err
+        }
+				return nil
+			}
+    }
+
+    return nil
+}
+
+func downloadFromUrl(url string) string {
+	tokens := strings.Split(url, "/")
+	fileName := filepath.Join(os.Getenv("GOPATH"), tokens[len(tokens)-1])
+	fmt.Println("Downloading", url, "to", fileName)
+
+	output, err := os.Create(fileName)
+	if err != nil {
+		fmt.Println("Error while creating", fileName, "-", err)
+		return ""
+	}
+	defer output.Close()
+
+	response, err := http.Get(url)
+	if err != nil {
+		fmt.Println("Error while downloading", url, "-", err)
+		return ""
+	}
+	defer response.Body.Close()
+
+	n, err := io.Copy(output, response.Body)
+	if err != nil {
+		fmt.Println("Error while downloading", url, "-", err)
+		return ""
+	}
+
+	fmt.Println(n, "bytes downloaded.")
+	return fileName
+}
+
+func join(separator string, args ...string) string {
+	return strings.Join(args, separator)
+}
+
+func installProtoc() {
+		if_not_exists("protoc", func() { // since we require a specific version anyway, wouldn't it be better to include this binary in the initial install filesystem?
+			const GOARCH string = runtime.GOARCH
+			const GOOS string = runtime.GOOS
+			if GOOS == "linux" && GOARCH == "amd64" { // check os and arch. If 64-bit linux, download binary. Other cases should be added as we start supporting other oses/architectures.
+				protocVersionNumber := strings.Split(ProtocVersion, " ")[1]
+				filename := downloadFromUrl(join("", "https://github.com/google/protobuf/releases/download/v", protocVersionNumber, "/protoc-", protocVersionNumber, "-linux-x86_64.zip"))
```

Comment:
I agree completely. I commented about it in the code at the top. I think that is definitely the better solution. We should have all the binaries for the different oses and architectures and just choose the right one when installing. But I don't know how our vendoring system works. I would just include it in the package directly...but then again, I would just have the whole thing in a package to begin with rather than run a build script..